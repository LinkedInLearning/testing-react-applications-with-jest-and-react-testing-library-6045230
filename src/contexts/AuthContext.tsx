import React, { createContext, useContext, useEffect, useState } from 'react';
import { jwtDecode } from 'jwt-decode';

interface User {
  id: number;
  email: string;
  username: string;
}

interface AuthContextType {
  user: User | null;
  loading: boolean;
  signIn: (email: string, password: string) => Promise<void>;
  signUp: (email: string, password: string, username: string) => Promise<void>;
  signOut: () => void;
}

const AuthContext = createContext<AuthContextType | undefined>(undefined);

const AUTH_API_URL = 'https://api.escuelajs.co/api/v1/auth';
const USER_API_URL = 'https://api.escuelajs.co/api/v1/users';

export function AuthProvider({ children }: { children: React.ReactNode }) {
  const [user, setUser] = useState<User | null>(null);
  const [loading, setLoading] = useState(true);

  useEffect(() => {
    const token = localStorage.getItem('token');
    if (token) {
      try {
        const decoded = jwtDecode<{ sub: string }>(token);
        // Fetch user details
        fetch(`${USER_API_URL}/${decoded.sub}`)
          .then(res => res.json())
          .then(userData => {
            setUser({
              id: userData.id,
              email: userData.email,
              username: userData.name
            });
          })
          .catch(() => {
            localStorage.removeItem('token');
          })
          .finally(() => {
            setLoading(false);
          });
      } catch {
        localStorage.removeItem('token');
        setLoading(false);
      }
    } else {
      setLoading(false);
    }
  }, []);

  const signIn = async (email: string, password: string) => {
    const response = await fetch(`${AUTH_API_URL}/login`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({ email, password }),
    });

    if (!response.ok) {
      throw new Error('Invalid credentials');
    }
   
    const { access_token } = await response.json();
    localStorage.setItem('token', access_token);
    
    const decoded = jwtDecode<{ sub: string }>(access_token);
    const userResponse = await fetch(`${USER_API_URL}/${decoded.sub}`);
    const userData = await userResponse.json();

    setUser({
      id: userData.id,
      email: userData.email,
      username: userData.name
    });
  };

  const signUp = async (email: string, password: string, username: string) => {
    const response = await fetch(USER_API_URL, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({
        name: username,
        email,
        password,
        avatar: 'https://picsum.photos/800'
      }),
    });

    if (!response.ok) {
      throw new Error('Registration failed');
    }

    await signIn(email, password);
  };

  const signOut = () => {
    localStorage.removeItem('token');
    setUser(null);
  };

  const value = {
    user,
    loading,
    signIn,
    signUp,
    signOut
  };

  return (
    <AuthContext.Provider value={value}>
      {!loading && children}
    </AuthContext.Provider>
  );
}

export function useAuth() {
  const context = useContext(AuthContext);
  if (context === undefined) {
    throw new Error('useAuth must be used within an AuthProvider');
  }
  return context;
}