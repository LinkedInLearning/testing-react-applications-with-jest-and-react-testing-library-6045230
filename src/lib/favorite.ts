import { PostWithCounts } from "../types/blog";

// Define the shape of favorites data in localStorage
type FavoritesStorage = Record<number, PostWithCounts>;

const addFavorite = (post: PostWithCounts) => {
  const favorites = getFavoritesRaw();
  favorites[post.id] = post;
  localStorage.setItem('favorites', JSON.stringify(favorites));
}

const removeFavorite = (postId: number) => {
  const favorites = getFavoritesRaw();
  delete favorites[postId];
  localStorage.setItem('favorites', JSON.stringify(favorites));
}

const getFavorites = () => {
  const favorites = getFavoritesRaw();
  return Object.values(favorites);
}

// Helper to safely parse localStorage data
const getFavoritesRaw = (): FavoritesStorage => {
  const storedFavorites = localStorage.getItem('favorites');
  try {
    return storedFavorites ? (JSON.parse(storedFavorites)) as FavoritesStorage : {};
  } catch (error) {
    console.error('Failed to parse favorites', error);
    return {};
  }
};

 const isFavorite = (postId: number) => {
  const favorites = getFavoritesRaw();
  return !!favorites[postId];
}

export {addFavorite, removeFavorite, getFavorites, isFavorite}
