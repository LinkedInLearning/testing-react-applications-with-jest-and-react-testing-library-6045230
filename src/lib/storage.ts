import type { LocalStorage } from '../types/blog';

const STORAGE_KEY = 'blog_data';

export function getStorageData(): LocalStorage {
  const data = localStorage.getItem(STORAGE_KEY);
  if (!data) {
    return { likes: [] };
  }
  return JSON.parse(data);
}

export function setStorageData(data: LocalStorage): void {
  localStorage.setItem(STORAGE_KEY, JSON.stringify(data));
}

export function toggleLike(postId: number): boolean {
  const data = getStorageData();
  const isLiked = data.likes.includes(postId);
  
  if (isLiked) {
    data.likes = data.likes.filter(id => id !== postId);
  } else {
    data.likes.push(postId);
  }
  
  setStorageData(data);
  return !isLiked;
}

export function isPostLiked(postId: number): boolean {
  const data = getStorageData();
  return data.likes.includes(postId);
}