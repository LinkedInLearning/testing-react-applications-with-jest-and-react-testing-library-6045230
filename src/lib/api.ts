import { getStorageData, isPostLiked } from './storage';
import type { Post, Comment, User, PostWithCounts } from '../types/blog';

const API_BASE_URL = 'https://jsonplaceholder.typicode.com';

const getAuthHeaders = () => {
  const token = localStorage.getItem('token');
  return token ? { Authorization: `Bearer ${token}` } : {};
};

export async function fetchPosts(query?: string): Promise<PostWithCounts[]> {
  const [posts, comments] = await Promise.all([
    fetch(`${API_BASE_URL}/posts`).then(res => res.json()),
    fetch(`${API_BASE_URL}/comments`).then(res => res.json())
  ]);
  
  const commentCounts = comments.reduce((acc: Record<number, number>, comment: Comment) => {
    acc[comment.postId] = (acc[comment.postId] || 0) + 1;
    return acc;
  }, {});

  const { likes } = getStorageData();
  
  const postsWithCounts = posts.map((post: Post) => ({
    ...post,
    commentsCount: commentCounts[post.id] || 0,
    likesCount: likes.includes(post.id) ? 1 : 0,
    isLiked: likes.includes(post.id)
  }));
  
  if (query) {
    return postsWithCounts.filter((post: PostWithCounts) => 
      post.title.toLowerCase().includes(query.toLowerCase()) ||
      post.body.toLowerCase().includes(query.toLowerCase())
    );
  }
  
  return postsWithCounts;
}

export async function fetchPost(id: number): Promise<PostWithCounts> {
  const [post, comments] = await Promise.all([
    fetch(`${API_BASE_URL}/posts/${id}`).then(res => res.json()),
    fetch(`${API_BASE_URL}/posts/${id}/comments`).then(res => res.json())
  ]);

  return {
    ...post,
    commentsCount: comments.length,
    likesCount: isPostLiked(post.id) ? 1 : 0,
    isLiked: isPostLiked(post.id)
  };
}

export async function fetchComments(postId: number): Promise<Comment[]> {
  const response = await fetch(`${API_BASE_URL}/posts/${postId}/comments`);
  return response.json();
}

export async function fetchUser(id: number): Promise<User> {
  const response = await fetch(`${API_BASE_URL}/users/${id}`);
  return response.json();
}

export async function createComment(postId: number, comment: Partial<Comment>): Promise<Comment> {
  const response = await fetch(`${API_BASE_URL}/comments`, {
    method: 'POST',
    headers: {
      'Content-type': 'application/json; charset=UTF-8',
      ...getAuthHeaders()
    },
    body: JSON.stringify({
      postId,
      ...comment,
    }),
  });
  return response.json();
}