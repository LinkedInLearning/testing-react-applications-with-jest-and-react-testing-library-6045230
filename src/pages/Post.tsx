import React, { useState, useEffect } from 'react';
import { useParams, useNavigate } from 'react-router-dom';
import { MessageSquare, ThumbsUp, Calendar, User } from 'lucide-react';
import styled from 'styled-components';
import { fetchPost, fetchComments, fetchUser, createComment } from '../lib/api';
import { toggleLike } from '../lib/storage';
import { CommentForm } from '../components/CommentForm';
import { Comment } from '../components/Comment';
import { useAuth } from '../contexts/AuthContext';
import type { PostWithCounts, Comment as CommentType, User as UserType } from '../types/blog';

const Container = styled.div`
  max-width: 64rem;
  margin: 0 auto;
`;

const Article = styled.article`
  background: white;
  border-radius: 1rem;
  overflow: hidden;
  box-shadow: 0 4px 6px -1px rgba(0, 0, 0, 0.1), 0 2px 4px -1px rgba(0, 0, 0, 0.06);
  margin-bottom: 2rem;
`;

const PostImage = styled.img`
  width: 100%;
  height: 24rem;
  object-fit: cover;
`;

const PostContent = styled.div`
  padding: 2.5rem;
`;

const PostTitle = styled.h1`
  font-size: 2.5rem;
  font-weight: bold;
  color: #111827;
  margin: 0 0 1.5rem 0;
  line-height: 1.2;
`;

const PostMeta = styled.div`
  display: flex;
  justify-content: space-between;
  align-items: center;
  padding: 1rem 0;
  border-top: 1px solid #e5e7eb;
  border-bottom: 1px solid #e5e7eb;
  margin-bottom: 2rem;
`;

const AuthorInfo = styled.div`
  display: flex;
  align-items: center;
  gap: 0.75rem;
  color: #4b5563;
  font-size: 0.875rem;

  span {
    display: flex;
    align-items: center;
    gap: 0.5rem;
  }
`;

const PostStats = styled.div`
  display: flex;
  gap: 1.5rem;
`;

const StatButton = styled.button<{ isLiked?: boolean }>`
  display: flex;
  align-items: center;
  gap: 0.5rem;
  color: ${props => props.isLiked ? '#2563eb' : '#6b7280'};
  background: none;
  border: none;
  cursor: pointer;
  padding: 0.5rem;
  border-radius: 0.5rem;
  transition: all 0.2s;

  &:hover {
    background-color: #f3f4f6;
  }
`;

const PostBody = styled.div`
  color: #374151;
  line-height: 1.8;
  font-size: 1.125rem;
  
  p {
    margin-bottom: 1.5rem;
  }
`;

const CommentsSection = styled.div`
  margin-top: 3rem;
`;

const CommentsHeader = styled.h2`
  font-size: 1.875rem;
  font-weight: bold;
  color: #111827;
  margin-bottom: 1.5rem;
`;

const CommentsList = styled.div`
  display: flex;
  flex-direction: column;
  gap: 1.5rem;
`;

const fetchWithRetry = async (fn: () => Promise<any>, retries = 3, delay = 200): Promise<any> => {
  for (let i = 0; i < retries; i++) {
    try {
      console.log({i})
      return await fn();
    } catch (err) {
      if (i === retries - 1) throw err;
      await new Promise(res => setTimeout(res, delay));
    }
  }
};

export function Post() {
  const { id } = useParams<{ id: string }>();
  const navigate = useNavigate();
  const { user } = useAuth();
  const [post, setPost] = useState<PostWithCounts | null>(null);
  const [author, setAuthor] = useState<UserType | null>(null);
  const [comments, setComments] = useState<CommentType[]>([]);
  const [loading, setLoading] = useState(true);

  useEffect(() => {
    if (!user) {
      navigate('/login');
      return;
    }

    const loadPost = async () => {
      if (!id) return;
      
      try {
        const postData = await fetchPost(parseInt(id));
        setPost(postData);

        const [authorData, commentsData] = await Promise.all([
          fetchUser(postData.userId),
          fetchWithRetry(() => fetchComments(postData.id), 3),
        ]);
       
        setAuthor(authorData);
        setComments(commentsData.reverse()); // Show newest comments first      
      } catch (error) {
        console.error('Error fetching post:', error);
        navigate('/');
      } finally {
        setLoading(false);
      }
    };

    loadPost();
  }, [id, navigate, user]);

  const handleComment = async ({ content }: { content: string }) => {
    if (!post || !user) return;

    try {
      const comment = await createComment(post.id, {
        name: user.username,
        email: user.email,
        body: content
      });
      
      setComments(prev => [comment, ...prev]); // Add new comment at the top
      setPost(prev => prev ? { ...prev, commentsCount: prev.commentsCount + 1 } : null);
    } catch (error) {
      console.error('Error adding comment:', error);
    }
  };

  const handleEditComment = async (commentId: number, content: string) => {
    setComments(comments.map(comment =>
      comment.id === commentId
        ? { ...comment, body: content }
        : comment
    ));
  };

  const handleDeleteComment = async (commentId: number) => {
    setComments(comments.filter(comment => comment.id !== commentId));
    setPost(prev => prev ? { ...prev, commentsCount: prev.commentsCount - 1 } : null);
  };

  const handleLike = () => {
    if (!post) return;
    
    const isLiked = toggleLike(post.id);
    setPost({ ...post, isLiked, likesCount: isLiked ? 1 : 0 });
  };

  if (loading) {
    return <div className="text-center py-8">Loading...</div>;
  }

  if (!post || !author) {
    return <div className="text-center py-8">Post not found</div>;
  }
  
  return (
    <Container>
      <Article>
        <PostImage
          src={`https://picsum.photos/seed/${post.id}/1200/600`}
          alt={post.title}
        />
        <PostContent>
          <PostTitle>{post.title}</PostTitle>
          <PostMeta>
            <AuthorInfo>
              <span>
                <User size={18} />
                {author.name}
              </span>
              <span>
                <Calendar size={18} />
                {new Date().toLocaleDateString()}
              </span>
            </AuthorInfo>
            <PostStats>
              <StatButton onClick={handleLike} isLiked={post.isLiked} data-testid={`likeButton-${post.id}`}>
                <ThumbsUp size={20} />
                <span>{post.likesCount}</span>
              </StatButton>
              <StatButton as="div">
                <MessageSquare size={20} />
                <span>{post.commentsCount}</span>
              </StatButton>
            </PostStats>
          </PostMeta>
          <PostBody>
            {post.body.split('\n').map((paragraph, index) => (
              <p key={index}>{paragraph}</p>
            ))}
          </PostBody>
        </PostContent>
      </Article>

      <CommentsSection>
        <CommentsHeader>Comments ({post.commentsCount})</CommentsHeader>
          <CommentForm onSubmit={handleComment} />
        <CommentsList>
          {comments.map((comment) => (
            <Comment
              key={comment.id}
              comment={comment}
              onEdit={handleEditComment}
              onDelete={handleDeleteComment}
              canModify={user?.email === comment.email}
            />
          ))}
        </CommentsList>
      </CommentsSection>
    </Container>
  );
}