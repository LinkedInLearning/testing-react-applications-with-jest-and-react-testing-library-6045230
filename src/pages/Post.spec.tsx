import { render, screen, waitFor } from '@testing-library/react';
import { describe, it, expect, vi, beforeEach, Mock } from 'vitest';

import { Post } from './Post';
import { MemoryRouter, Route, Routes } from 'react-router-dom';

import * as api from '../lib/api';
import * as storage from '../lib/storage';
import { useAuth } from '../contexts/AuthContext';
import userEvent from '@testing-library/user-event';

vi.mock('../lib/api');
vi.mock('../lib/storage');
vi.mock('../contexts/AuthContext', () => ({
  useAuth: vi.fn(),
}));

const mockNavigate = vi.fn();
vi.mock('react-router-dom', async () => {
  const actual = await vi.importActual<typeof import('react-router-dom')>('react-router-dom');
  return {
    ...actual,
    useParams: () => ({ id: '1' }),
    useNavigate: () => mockNavigate,
  };
});

const mockPost = {
  id: 1,
  title: 'Test Post',
  body: 'This is a test post.\nSecond paragraph.',
  userId: 1,
  commentsCount: 1,
  likesCount: 0,
  isliked: false,
};

const mockUser = {
  id: 1,
  name: 'Jane Doe',
  username: 'janedoe',
  email: 'jane@example.com',
};

const mockComments = [
  {
    id: 1,
    name: 'Commenter',
    email: 'jane@example.com',
    body: 'Nice post!',
    postId: '1'
  }
];

const setup = (authUser = mockUser) => {
  (useAuth as Mock).mockReturnValue({ user: authUser });

  (api.fetchPost as Mock).mockResolvedValue(mockPost);
  (api.fetchUser as Mock).mockResolvedValue(mockUser);
  (api.fetchComments as Mock).mockResolvedValue(mockComments);
};

describe('<Post />', () => {
  beforeEach(() => {
    setup()
    vi.clearAllMocks();

  });

  const renderComponent = () => {
    return (
      render(
        <MemoryRouter initialEntries={['/posts/1']}>
          <Routes>
            <Route path="/posts/:id" element={<Post />} />
          </Routes>
        </MemoryRouter>
      )
    )
  }

  it('should match snapshot when fully loaded', async () => {
    const { asFragment, findByText } = renderComponent();
    
    // Wait for the component to fully load
    await findByText('Test Post');
    
    // Take a snapshot of the fully rendered component
    expect(asFragment()).toMatchSnapshot();
  });

  it('should match snapshot when post is not found', async () => {
    // Mock post not found
    (api.fetchPost as Mock).mockResolvedValue(null);
    
    const { asFragment, findByText } = renderComponent();
    
    // Wait for the not found message to appear
    await findByText('Post not found');
    
    // Take a snapshot of the not found state
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders loading state', async () => {
    renderComponent()

    expect(screen.getByText(/loading/i)).toBeInTheDocument();
    await waitFor(() => expect(screen.queryByText(/loading/i)).not.toBeInTheDocument());
  });

  it('redirects if user is not logged in', async () => {
    (useAuth as Mock).mockReturnValue({ user: null });
    (api.fetchUser as Mock).mockResolvedValue(null);
    renderComponent()

    await waitFor(() => {
      expect(mockNavigate).toHaveBeenCalledWith('/login');
    });
  });

  it('handles post not found', async () => {
    (api.fetchPost as Mock).mockResolvedValue(null);

    renderComponent()

    await waitFor(() => {
      expect(screen.getByText(/post not found/i)).toBeInTheDocument();
    });
  });

  it('handles like button toggle', async () => {
    (storage.toggleLike as Mock).mockReturnValue(true);
    const user = userEvent.setup()
    renderComponent()

    expect(await screen.findByText('Test Post')).toBeInTheDocument();

    const likeButton = await screen.getAllByTestId('likeButton-1')[0]
    await user.click(likeButton);
    expect(storage.toggleLike).toHaveBeenCalledWith(1);
  });

  it('renders post content and comments', async () => {
    renderComponent()

    expect(await screen.findByText('Test Post')).toBeInTheDocument();
    expect(screen.getByText('Jane Doe')).toBeInTheDocument();
    expect(screen.getByText('Nice post!')).toBeInTheDocument();
    expect(screen.getByText(/comments \(1\)/i)).toBeInTheDocument();
  });

  it('adds a new comment on submission', async () => {
    (api.createComment as Mock).mockResolvedValue({
      id: 2,
      name: 'Jane Doe',
      email: 'jane@example.com',
      body: 'New comment',
    });

    const user = userEvent.setup()
    renderComponent()

    const textarea = await screen.findByRole('textbox');
    // fireEvent.change(textarea, { target: { value: 'New comment' } });
    await user.type(textarea, 'New comment');

    const button = screen.getByRole('button', { name: /comment/i });
    // fireEvent.click(button);
    await user.click(button);

    await waitFor(() => {
      expect(screen.getByText('New comment')).toBeInTheDocument();
      expect(screen.getByText(/comments \(2\)/i)).toBeInTheDocument();
    });
  });

  it('handles comment creation error gracefully', async () => {
    const error = new Error('Failed to comment');
    (api.createComment as Mock).mockRejectedValue(error);
    const user = userEvent.setup()
    const consoleErrorSpy = vi.spyOn(console, 'error').mockImplementation(() => { });

    renderComponent()

    const textarea = await screen.findByRole('textbox');
    await user.type(textarea, 'Will fail');

    const button = screen.getByRole('button', { name: /comment/i });
    await user.click(button);

    await waitFor(() => {
      expect(consoleErrorSpy).toHaveBeenCalledWith('Error adding comment:', error);
    });

    consoleErrorSpy.mockRestore();
  });

  it('deletes a comment', async () => {
    vi.spyOn(window, 'confirm').mockReturnValue(true);
    (api.fetchComments as Mock).mockResolvedValue([
      ...mockComments,
      {
        id: 2,
        name: 'Commenter',
        email: 'jane@example.com',
        body: 'Well said',
        postId: '1'
      }
    ]);
    const user = userEvent.setup();
    await renderComponent()


    expect(await screen.findByText('Test Post')).toBeInTheDocument();
    await user.click(screen.getByTestId(`${mockComments[0].id}-deleteBtn`))

    expect(screen.queryByText(mockComments[0].body)).not.toBeInTheDocument();

  });

  it('updates comment content when edited', async () => {
    const updatedText = 'Update comment';
    (api.createComment as Mock).mockResolvedValue({
      ...mockComments[0],
      body: updatedText,
    });
    const user = userEvent.setup();
    await renderComponent()


    expect(await screen.findByText('Test Post')).toBeInTheDocument();
    await user.click(screen.getByTestId(`${mockComments[0].id}-editBtn`))

    const textarea = await screen.getByTestId('editInput');
    await user.clear(textarea);
    await user.type(textarea, updatedText);

    const button = screen.getByRole('button', { name: /save changes/i });
    await user.click(button);

    await waitFor(() => {
      expect(screen.getByText(updatedText)).toBeInTheDocument();
      expect(screen.getByText(/comments \(1\)/i)).toBeInTheDocument();
    });
  });

  it('handles rapid comment edits correctly', async () => {
    // Create mock with delayed resolution
    let resolveFirstEdit!: (value: any) => void;
    let resolveSecondEdit!: (value: any) => void;

    const firstEditPromise = new Promise(resolve => {
      resolveFirstEdit = resolve;
    });

    const secondEditPromise = new Promise(resolve => {
      resolveSecondEdit = resolve;
    });

    // First edit will resolve after second edit
    (api.createComment as Mock)
      .mockImplementationOnce(() => firstEditPromise)
      .mockImplementationOnce(() => secondEditPromise);

    const user = userEvent.setup();
    await renderComponent();


    expect(await screen.findByText('Test Post')).toBeInTheDocument();
    // Start first edit
    await user.click(screen.getByTestId(`${mockComments[0].id}-editBtn`));
    const textarea = await screen.getByTestId('editInput');
    await user.clear(textarea);
    await user.type(textarea, 'First edit');
    await user.click(screen.getByRole('button', { name: /save changes/i }));

    // Start second edit before first completes
    await user.click(screen.getByTestId(`${mockComments[0].id}-editBtn`));
    const textareaAgain = await screen.getByTestId('editInput');
    await user.clear(textareaAgain);
    await user.type(textareaAgain, 'Second edit');
    await user.click(screen.getByRole('button', { name: /save changes/i }));

    // Resolve second edit first
    resolveSecondEdit({
      ...mockComments[0],
      body: 'Second edit'
    });

    // Then resolve first edit
    resolveFirstEdit({
      ...mockComments[0],
      body: 'First edit'
    });

    // Final state should show the second edit
    await waitFor(() => {
      expect(screen.getByText('Second edit')).toBeInTheDocument();
      expect(screen.queryByText('First edit')).not.toBeInTheDocument();
    });
  });

  it('cancels pending requests when unmounting', async () => {
    // Create a mock that never resolves during the test
    const neverResolvingPromise = new Promise(() => { });
    (api.fetchComments as Mock).mockReturnValue(neverResolvingPromise);

    // Mock console.error to catch "Can't perform a React state update on an unmounted component" warnings
    const consoleErrorSpy = vi.spyOn(console, 'error').mockImplementation(() => { });

    const { unmount } = render(
      <MemoryRouter initialEntries={['/posts/1']}>
        <Routes>
          <Route path="/posts/:id" element={<Post />} />
        </Routes>
      </MemoryRouter>
    );

    // Wait for component to start rendering
    await screen.findByText(/loading/i);

    // Unmount component while request is still pending
    unmount();

    // Verify no console errors about updating unmounted components
    expect(consoleErrorSpy).not.toHaveBeenCalled();

    consoleErrorSpy.mockRestore();
  });

  it('retries failed comment fetch up to 3 times', async () => {
    const error = new Error('Network error');

    // Fail twice, succeed on third try
    (api.fetchComments as Mock)
      .mockRejectedValueOnce(error)
      .mockRejectedValueOnce(error)
      .mockResolvedValueOnce([{
        ...mockComments[0],
        body: 'Comment after retry',
      }]);

    renderComponent();

    // Initially shows loading
    expect(screen.getByText(/loading/i)).toBeInTheDocument();

    // Eventually shows the comment that succeeded after retries
    expect(await screen.findByText(/comment after retry/i)).toBeInTheDocument();

    // Verify fetchComments was called exactly 3 times
    await waitFor(() => expect(api.fetchComments).toHaveBeenCalledTimes(3))
  });
});
