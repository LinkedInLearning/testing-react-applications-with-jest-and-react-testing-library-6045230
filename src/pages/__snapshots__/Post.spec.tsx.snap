// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<Post /> > should match snapshot when fully loaded 1`] = `
<DocumentFragment>
  <div
    class="sc-cRHGzb hVXCSc"
  >
    <article
      class="sc-jOTjWM hQGDsi"
    >
      <img
        alt="Test Post"
        class="sc-cOPLXt GHRBo"
        src="https://picsum.photos/seed/1/1200/600"
      />
      <div
        class="sc-jXuTBK bbmEcz"
      >
        <h1
          class="sc-cXrvCr eZardT"
        >
          Test Post
        </h1>
        <div
          class="sc-jUCZac jREzfA"
        >
          <div
            class="sc-cUzBaJ GobPe"
          >
            <span>
              <svg
                class="lucide lucide-user "
                fill="none"
                height="18"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="18"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M19 21v-2a4 4 0 0 0-4-4H9a4 4 0 0 0-4 4v2"
                />
                <circle
                  cx="12"
                  cy="7"
                  r="4"
                />
              </svg>
              Jane Doe
            </span>
            <span>
              <svg
                class="lucide lucide-calendar "
                fill="none"
                height="18"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="18"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M8 2v4"
                />
                <path
                  d="M16 2v4"
                />
                <rect
                  height="18"
                  rx="2"
                  width="18"
                  x="3"
                  y="4"
                />
                <path
                  d="M3 10h18"
                />
              </svg>
              5/3/2025
            </span>
          </div>
          <div
            class="sc-lnQFQe ilKtLK"
          >
            <button
              class="sc-enNhQL dQDrps"
              data-testid="likeButton-1"
            >
              <svg
                class="lucide lucide-thumbs-up "
                fill="none"
                height="20"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="20"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M7 10v12"
                />
                <path
                  d="M15 5.88 14 10h5.83a2 2 0 0 1 1.92 2.56l-2.33 8A2 2 0 0 1 17.5 22H4a2 2 0 0 1-2-2v-8a2 2 0 0 1 2-2h2.76a2 2 0 0 0 1.79-1.11L12 2h0a3.13 3.13 0 0 1 3 3.88Z"
                />
              </svg>
              <span>
                0
              </span>
            </button>
            <div
              class="sc-enNhQL dQDrps"
            >
              <svg
                class="lucide lucide-message-square "
                fill="none"
                height="20"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="20"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M21 15a2 2 0 0 1-2 2H7l-4 4V5a2 2 0 0 1 2-2h14a2 2 0 0 1 2 2z"
                />
              </svg>
              <span>
                1
              </span>
            </div>
          </div>
        </div>
        <div
          class="sc-dUwGTt kLOsqR"
        >
          <p>
            This is a test post.
          </p>
          <p>
            Second paragraph.
          </p>
        </div>
      </div>
    </article>
    <div
      class="sc-eUzcTs cCNSRk"
    >
      <h2
        class="sc-gKcHA-D iCzahW"
      >
        Comments (1)
      </h2>
      <form
        class="sc-VHjGu fXXijF"
      >
        <textarea
          class="sc-gnOvAp ioKTKF"
          data-testid="createInput"
          name="content"
          placeholder="Write your comment..."
        />
        <button
          class="sc-glJfXD hiMLPh"
          type="submit"
        >
          Post Comment
        </button>
      </form>
      <div
        class="sc-hKfdAC lgSTfe"
      >
        <div
          class="sc-dOfePm czXTTb"
        >
          <div
            class="sc-jgmqJh katOQV"
          >
            <div
              class="sc-gIIpAQ icdtER"
            >
              <h4>
                Commenter
              </h4>
              <p>
                <svg
                  class="lucide lucide-clock "
                  fill="none"
                  height="14"
                  stroke="currentColor"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="2"
                  viewBox="0 0 24 24"
                  width="14"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <circle
                    cx="12"
                    cy="12"
                    r="10"
                  />
                  <polyline
                    points="12 6 12 12 16 14"
                  />
                </svg>
                5/3/2025
              </p>
            </div>
            <div
              class="sc-kFkjun SGQqy"
            >
              <button
                class="sc-ihGilW dltwXA"
                data-testid="1-editBtn"
                title="Edit"
              >
                <svg
                  class="lucide lucide-pencil "
                  fill="none"
                  height="18"
                  stroke="currentColor"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="2"
                  viewBox="0 0 24 24"
                  width="18"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M17 3a2.85 2.83 0 1 1 4 4L7.5 20.5 2 22l1.5-5.5Z"
                  />
                  <path
                    d="m15 5 4 4"
                  />
                </svg>
              </button>
              <button
                class="sc-ihGilW dltwXA"
                data-testid="1-deleteBtn"
                title="Delete"
              >
                <svg
                  class="lucide lucide-trash2 "
                  fill="none"
                  height="18"
                  stroke="currentColor"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="2"
                  viewBox="0 0 24 24"
                  width="18"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M3 6h18"
                  />
                  <path
                    d="M19 6v14c0 1-1 2-2 2H7c-1 0-2-1-2-2V6"
                  />
                  <path
                    d="M8 6V4c0-1 1-2 2-2h4c1 0 2 1 2 2v2"
                  />
                  <line
                    x1="10"
                    x2="10"
                    y1="11"
                    y2="17"
                  />
                  <line
                    x1="14"
                    x2="14"
                    y1="11"
                    y2="17"
                  />
                </svg>
              </button>
            </div>
          </div>
          <p
            class="sc-jRLeyu dRNHGZ"
          >
            Nice post!
          </p>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`<Post /> > should match snapshot when post is not found 1`] = `
<DocumentFragment>
  <div
    class="text-center py-8"
  >
    Post not found
  </div>
</DocumentFragment>
`;
